crates/bevy_asset/src/path.rs

```diff
@@ -659,12 +672,80 @@ pub(crate) fn normalize_path(path: &Path) -> PathBuf {
     result_path
 }
 
+/// Normalizes the path by collapsing all occurrences of '.' and '..' dot-segments where possible
+/// as per [RFC 1808](https://datatracker.ietf.org/doc/html/rfc1808)
+/// Returns a borrowed path if no normalization was necessary, otherwise returns an owned normalized path.
+pub(crate) fn maybe_normalize_path<'a, P: AsRef<Path> + Into<CowArc<'a, Path>> + 'a>(
+    as_path: P,
+) -> CowArc<'a, Path> {
+    let path = as_path.as_ref();
+    let mut result_path: core::cell::OnceCell<PathBuf> = core::cell::OnceCell::new();
+    let init = |i: usize| -> PathBuf { path.iter().take(i).collect() };
+
+    for (i, elt) in path.iter().enumerate() {
+        if elt == "." {
+            result_path.get_or_init(|| init(i));
+        } else if elt == ".." {
+            result_path.get_or_init(|| init(i));
+
+            if let Some(path) = result_path.get_mut()
+                && !path.pop()
+            {
+                // Preserve ".." if insufficient matches (per RFC 1808).
+                path.push(elt);
+            }
+        } else if let Some(path) = result_path.get_mut() {
+            path.push(elt);
+        }
+    }
+
+    match result_path.into_inner() {
+        Some(path_buf) => CowArc::Owned(path_buf.into()),
+        None => as_path.into(),
+    }
+}
+
+trait NormalizeCowPath {
```

Comment:
I don't like this magic. We only have a few callsites and this isn't pub anyway, so I'd prefer just making this a function that we call on our path instead of a trait.